packages/squiggle-lang/src/value/domain.ts

```diff
@@ -0,0 +1,55 @@
+import { ErrorMessage, REOther } from "../errors/messages.js";
+import { Err, Ok, result } from "../utility/result.js";
+import { Value } from "./index.js";
+
+abstract class BaseDomain {
+  abstract type: string;
+
+  abstract toString(): string;
+
+  abstract includes(value: Value): boolean;
+}
+
+class NumericRangeDomain extends BaseDomain {
+  readonly type = "NumericRange";
+
+  constructor(public min: number, public max: number) {
+    super();
+  }
+
+  toString() {
+    return `Range(${this.min} to ${this.max})`;
+  }
+
+  includes(value: Value) {
+    return (
+      value.type === "Number" &&
+      value.value >= this.min &&
+      value.value <= this.max
+    );
+  }
+}
+
+export type Domain = NumericRangeDomain;
+
+export function annotationToDomain(value: Value): result<Domain, ErrorMessage> {
+  if (value.type !== "Array") {
+    return Err(new REOther("Only array domains are supported"));
+  }
+  if (value.value.length !== 2) {
+    return Err(new REOther("Expected two-value array"));
+  }
+  const [min, max] = value.value;
+  if (min.type !== "Number") {
+    return Err(new REOther("Min value is not a number"));
```

Comment:
Maybe, `Domain min must be a number, was given a ${min.type}.`

Reply:
I think this entire function (`annotationToDomain`) should be in stdlib and use `frType` helpers to unpack its values.

So the entire syntax sugar for transforming `[number, number]` to domain objects is that all annotations are wrapped in `Domain.create(...)`. `Domain.create` would be polymorphic on various value types; `Domain.create(domain)` would be an identity function.

It might make these error messages slightly worse, but it will be more scalable as we add more domains.

Reply:
Makes sense, I like that that sounds generic+scalable.